pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,349 @@
+		Endpoint:        os.Getenv("ALIYUNOSS_ENDPOINT"),
+		Bucket:          os.Getenv("ALIYUNOSS_BUCKET"),
+		AccessKeyID:     os.Getenv("ALIYUNOSS_ACCESS_KEY_ID"),
+		AccessKeySecret: os.Getenv("ALIYUNOSS_ACCESS_KEY_SECRET"),
+	}
+
+	if c.Endpoint == "" || c.AccessKeyID == "" || c.AccessKeySecret == "" {
+		return nil, nil, errors.New("aliyun oss endpoint or access_key_id or access_key_secret " +
+			"is not present in config file")
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "true" {
+		t.Log("ALIYUNOSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset ALIYUNOSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+		return NewTestBucketFromConfig(t, c, true)
+	}
+	return NewTestBucketFromConfig(t, c, false)
+}
+
+func calculateChunks(name string, r io.Reader) (int, int64, error) {
+	switch r.(type) {
+	case *os.File:
+		f, _ := r.(*os.File)
+		if fileInfo, err := f.Stat(); err == nil {
+			s := fileInfo.Size()
+			return int(math.Floor(float64(s) / PartSize)), s % PartSize, nil
+		}
+	case *strings.Reader:
+		f, _ := r.(*strings.Reader)
+		return int(math.Floor(float64(f.Size()) / PartSize)), f.Size() % PartSize, nil
+	}
+	return -1, 0, errors.New("unsupported implement of io.Reader")
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	chunksnum, lastslice, err := calculateChunks(name, r)
+	if err != nil {
+		return err
+	}
+
+	ncloser := ioutil.NopCloser(r)
+	switch chunksnum {
+	case 0:
+		if err := b.bucket.PutObject(name, ncloser); err != nil {
+			return errors.Wrap(err, "failed to upload oss object")
+		}
+	default:
+		{
+			init, err := b.bucket.InitiateMultipartUpload(name)
+			if err != nil {
+				return errors.Wrap(err, "failed to initiate multi-part upload")
+			}
+			chunk := 0
+			uploadEveryPart := func(everypartsize int64, cnk int) (alioss.UploadPart, error) {
+				prt, err := b.bucket.UploadPart(init, ncloser, everypartsize, cnk)
+				if err != nil {
+					if err := b.bucket.AbortMultipartUpload(init); err != nil {
+						return prt, errors.Wrap(err, "failed to abort multi-part upload")
+					}
+
+					return prt, errors.Wrap(err, "failed to upload multi-part chunk")
+				}
+				return prt, nil
+			}
+			var parts []alioss.UploadPart
+			for ; chunk < chunksnum; chunk++ {
+				part, err := uploadEveryPart(PartSize, chunk+1)
+				if err != nil {
+					return errors.Wrap(err, "failed to upload every part")
+				}
+				parts = append(parts, part)
+			}
+			if lastslice != 0 {
+				part, err := uploadEveryPart(lastslice, chunksnum+1)
+				if err != nil {
+					return errors.Wrap(err, "failed to upload the last chunk")
+				}
+				parts = append(parts, part)
+			}
+			_, err = b.bucket.CompleteMultipartUpload(init, parts)
```

Comment:
we can use `if err := ...; err != nil` pattern here (:

Reply:
OK